src/System.Slices/src/System/SpanEnumerators.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System
+{
+    public partial struct Span<T>
+    {
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// </summary>
+        public struct Enumerator
+        {
```

Comment:
In ideal world, JIT would allocate a class iterator on the stack when possible (it seems possible in typical foreach), then devirtualize call to Dispose and notice that Dispose is actually empty and thus erase it as well as redundant try/finally.

In such case we could just have one enumerator that is a class. This is not the case today. :-(